src/OpenTelemetry/Metrics/MetricProvider.cs

```diff
@@ -0,0 +1,147 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Threading;
+using System.Threading.Tasks;
+
+#nullable enable
+
+namespace OpenTelemetry.Metrics
+{
+    public class MetricProvider : IDisposable
+    {
+        private BuildOptions options;
+        private ConcurrentDictionary<Meter, int> meters;
+        private MeterListener listener;
+        private CancellationTokenSource cts;
+        private Task observerTask;
+
+        internal MetricProvider(BuildOptions options)
+        {
+            this.options = options;
+
+            this.meters = new ConcurrentDictionary<Meter, int>();
+
+            this.listener = new MeterListener()
+            {
+                InstrumentPublished = (instrument, listener) => this.InstrumentPublished(instrument, listener),
+                MeasurementsCompleted = (instrument, state) => this.MeasurementsCompleted(instrument, state),
+            };
+            this.listener.SetMeasurementEventCallback<double>((i, m, l, c) => this.MeasurementRecorded(i, m, l, c));
+            this.listener.SetMeasurementEventCallback<float>((i, m, l, c) => this.MeasurementRecorded(i, m, l, c));
+            this.listener.SetMeasurementEventCallback<long>((i, m, l, c) => this.MeasurementRecorded(i, m, l, c));
+            this.listener.SetMeasurementEventCallback<int>((i, m, l, c) => this.MeasurementRecorded(i, m, l, c));
+            this.listener.SetMeasurementEventCallback<short>((i, m, l, c) => this.MeasurementRecorded(i, m, l, c));
+            this.listener.SetMeasurementEventCallback<byte>((i, m, l, c) => this.MeasurementRecorded(i, m, l, c));
+
+            this.listener.Start();
+
+            this.cts = new CancellationTokenSource();
+
+            var token = this.cts.Token;
+            this.observerTask = Task.Run(async () =>
+            {
+                while (!token.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await Task.Delay(this.options.ObservationPeriodMilliseconds, token);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                    }
+
+                    this.listener.RecordObservableInstruments();
+                }
+            });
+        }
+
+        public Meter GetMeter(string name, string version)
+        {
+            var meter = new Meter(name, version);
```

Comment:
Might need some concurrency support here.

Reply:
For example, if the same name is already used, we could end up creating a Meter via .NET API but fail to update the `this.meters` dictionary.

If creating a dangling meter doesn't have any negative impact, it is probably fine.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/new_api.md#get-a-meter

_It is unspecified whether or under which conditions the same or different Meter instances are returned from this functions._

We might need to define the behavior for OpenTelemetry .NET?